### PR TITLE
pkg/checkpoint: support projected secret/configmap volumes

### LINKED ISSUE #969: pod-checkpointer doesn't handle a pod using projected volumes correctly
issue: our apiserver pod is created with kubeadm and contains projected volumes.

eg:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    checkpointer.alpha.coreos.com/checkpoint: "true"
    kubernetes.io/created-by: |
      {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"kube-system","name":"kube-apiserver-74d6c88bdc","uid":"c1d86535-38fb-11e8-b4cb-fa163e479b7b","apiVersion":"extensions","resourceVersion":"1246685578"}}
    scheduler.alpha.kubernetes.io/critical-pod: ""
  creationTimestamp: 2018-04-30T15:37:57Z
  generateName: kube-apiserver-74d6c88bdc-
  labels:
    app: kube-apiserver
    appId: kube-apiserver
    k8s-app: kube-apiserver
    pod-template-hash: "3082744687"
  name: kube-apiserver-74d6c88bdc-4szn7
  namespace: kube-system
  ownerReferences:
  - apiVersion: extensions/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: kube-apiserver-74d6c88bdc
    uid: c1d86535-38fb-11e8-b4cb-fa163e479b7b
  resourceVersion: "1246687617"
  selfLink: /api/v1/namespaces/kube-system/pods/kube-apiserver-74d6c88bdc-4szn7
  uid: 74c541bf-4c8c-11e8-886d-fa163e8fc7ed
spec:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: appId
            operator: In
            values:
            - kube-apiserver
        topologyKey: failure-domain.beta.kubernetes.io/zone
  containers:
  - command:
    - /hyperkube
    - apiserver
    - --apiserver-count=1
    - --bind-address=0.0.0.0
    - --secure-port=443
    - --insecure-port=8080
    - --advertise-address=$(POD_IP)
    - --etcd-servers=http://127.0.0.1:2379
    - --storage-backend=etcd2
    - --storage-media-type=application/json
    - --allow-privileged=true
    - --service-cluster-ip-range=172.31.128.0/22
    - --admission-control=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,AlwaysPullImages,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,ResourceQuota
    - --tls-cert-file=/etc/kubernetes/pki/apiserver.crt
    - --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
    - --service-account-key-file=/etc/kubernetes/pki/sa.pub
    - --client-ca-file=/etc/kubernetes/pki/ca.crt
    - --basic-auth-file=/etc/kubernetes/static-api-users/users-file
    - --token-auth-file=/etc/kubernetes/static-api-tokens/token-file
    - --runtime-config=batch/v2alpha1
    - --feature-gates=ExperimentalCriticalPodAnnotation=true,TaintBasedEvictions=true
    - --authorization-mode=Node,RBAC
    - --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt
    - --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key
    - --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
    - --requestheader-allowed-names=front-proxy-client
    - --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt
    - --enable-bootstrap-token-auth=true
    - --requestheader-username-headers=X-Remote-User
    - --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt
    - --requestheader-group-headers=X-Remote-Group
    - --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key
    env:
    - name: POD_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    image: quay.io/coreos/hyperkube:v1.8.7_coreos.0
    imagePullPolicy: Always
    name: kube-apiserver
    resources:
      limits:
        cpu: "2"
        memory: 4000Mi
      requests:
        cpu: "1"
        memory: 4000Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/ssl/certs
      name: ca-certs
      readOnly: true
    - mountPath: /etc/kubernetes/pki
      name: k8s-certs
      readOnly: true
    - mountPath: /etc/kubernetes/static-api-tokens
      name: static-api-tokens
      readOnly: true
    - mountPath: /etc/kubernetes/static-api-users
      name: static-api-users
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-6zri8
      readOnly: true
  dnsPolicy: ClusterFirstWithHostNet
  hostNetwork: true
  nodeName: kubernetes-cit-kubernetes-cr1-7-1524143077
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - key: CriticalAddonsOnly
    operator: Exists
  - effect: NoExecute
    key: node.alpha.kubernetes.io/notReady
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.alpha.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: k8s-certs
    projected:
      defaultMode: 420
      sources:
      - secret:
          items:
          - key: tls.crt
            path: ca.crt
          name: ca
      - secret:
          items:
          - key: tls.crt
            path: apiserver.crt
          - key: tls.key
            path: apiserver.key
          name: apiserver
      - secret:
          items:
          - key: tls.crt
            path: apiserver-kubelet-client.crt
          - key: tls.key
            path: apiserver-kubelet-client.key
          name: apiserver-kubelet-client
      - secret:
          items:
          - key: tls.crt
            path: sa.pub
          name: sa
      - secret:
          items:
          - key: tls.crt
            path: front-proxy-ca.crt
          name: front-proxy-ca
      - secret:
          items:
          - key: tls.crt
            path: front-proxy-client.crt
          - key: tls.key
            path: front-proxy-client.key
          name: front-proxy-client
  - hostPath:
      path: /etc/ssl/certs
      type: DirectoryOrCreate
    name: ca-certs
  - name: static-api-tokens
    secret:
      defaultMode: 420
      secretName: static-api-tokens
  - name: static-api-users
    secret:
      defaultMode: 420
      secretName: static-api-users
  - name: default-token-6zri8
    secret:
      defaultMode: 420
      secretName: default-token-6zri8
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-04-30T15:38:15Z
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-04-30T15:39:13Z
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-04-30T15:38:15Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://40bcece6dc20777f5f95cbaa0de4347b1d2dbe0eba2eb94c1bfd3396e3a3c54b
    image: quay.io/coreos/hyperkube:v1.8.7_coreos.0
    imageID: docker-pullable://quay.io/coreos/hyperkube@sha256:6d138fa463330786e22d0d2fef3d47c894036f701d884ced45dfbaf8d4895ff4
    lastState: {}
    name: kube-apiserver
    ready: true
    restartCount: 0
    state:
      running:
        startedAt: 2018-04-30T15:39:12Z
  hostIP: 10.33.255.83
  phase: Running
  podIP: 10.33.255.83
  qosClass: Burstable
  startTime: 2018-04-30T15:38:15Z
```

the checkpointed apiserver pod still has the k8s-cert volume unchanged so on apiserver failure startup fails due to not being able to pull secrets from the api

### PR DESCRIPTION
**Pull request by stuart-warren**
to allow kubeadm versions of kube-apiserver pod to still work

fixes #969

because secrets and configmaps currently have different base hostpaths, we can't work with projected volumes containing both.

### CODE REVIEW

File: pkg/checkpoint/secret.go [hunk 1]
```diff
@@ -19,13 +21,20 @@ func (c *checkpointer) checkpointSecretVolumes(pod *v1.Pod) (*v1.Pod, error) {
...
+				return nil, fmt.Errorf("failed to checkpoint secret for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+			}
 		}
 
-		_, err := c.checkpointSecret(pod.Namespace, pod.Name, v.Secret.SecretName, uid, gid)
-		if err != nil {
-			return nil, fmt.Errorf("failed to checkpoint secret for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		if v.VolumeSource.Projected != nil {
+			p := v.VolumeSource.Projected
+			err := c.checkpointProjectedSecrets(pod.Namespace, pod.Name, v.Name, p.Sources, uid, gid)
```
Comment (ericchiang) [inline pkg/checkpoint/secret.go]: This doesn't checkpoint the configmaps either?

I'd also prefer error'ing if we detect a downward API here. Better to be explicit then create hard to debug scenarios

File: pkg/checkpoint/secret.go [hunk 2]
```diff
@@ -50,14 +59,84 @@ func (c *checkpointer) checkpointSecret(namespace, podName, secretName string, u
...
+		if err := writeAndAtomicRename(filepath.Join(basePath, f), d, uid, gid, defaultMode); err != nil {
 			return "", fmt.Errorf("failed to write secret %s: %v", secret.Name, err)
 		}
 	}
 
 	return basePath, nil
 }
 
+// checkpointSecretProjection will locally store secret data from volumes with ProjectedVolumeSources.
+// The path to the secret data becomes: checkpointSecretPath/namespace/podname/volumename/secret.file
```
Comment (ericchiang) [inline pkg/checkpoint/secret.go]: The logic below is a little hard to follow, but is this right?

```
checkpointSecretPath/namespace/podname/volumename/secret.file
```

I would have expected a constructed folder here, not a secret file.

```
checkpointSecretPath/namespace/podname/volumename/( directory with projected secrets and configmaps )
```

Comment (stuart-warren) [reply pkg/checkpoint/secret.go]: the projected volume specifies the path that each key is written to which can include folder names. this is mentioned on the next line

https://kubernetes.io/docs/concepts/storage/volumes/#projected

File: pkg/checkpoint/secret.go (hunk 1 above)
Comment (stuart-warren) [reply pkg/checkpoint/secret.go]: added support for configmaps and error if using downward api

File: pkg/checkpoint/pod.go [hunk 3]
```diff
@@ -72,6 +72,15 @@ func sanitizeCheckpointPod(cp *v1.Pod) *v1.Pod {
 		} else if v.ConfigMap != nil {
 			v.HostPath = &v1.HostPathVolumeSource{Path: configMapPath(cp.Namespace, cp.Name, v.ConfigMap.Name)}
 			v.ConfigMap = nil
+		} else if v.Projected != nil {
+			// can only currently support projected volumes with all secrets or all configmaps
+			if volumeHasSecretProjection(v) {
+				v.HostPath = &v1.HostPathVolumeSource{Path: secretPath(cp.Namespace, cp.Name, v.Name)}
+				v.Projected = nil
+			} else if volumeHasConfigMapProjection(v) {
```
Comment (rphillips) [inline pkg/checkpoint/pod.go]: a projection could have a secret and a configmap, right? So this else if would be just an if statement.

Comment (stuart-warren) [reply pkg/checkpoint/pod.go]: please see update to pull request description

Comment (stuart-warren) [reply pkg/checkpoint/pod.go]: it's not something my work colleagues and I currently need but we would like some basic support for projected volumes hence the PR :)

Comment (ericchiang) [reply pkg/checkpoint/pod.go]: I would expect this to support a projected volume with both configmaps and secrets. It doesn't make sense to advertise this as "supporting projected volumes" otherwise.

Comment (stuart-warren) [reply pkg/checkpoint/pod.go]: OK, so I need to serialise secrets and config maps into the same directory for a volume.

Would you prefer all secrets and config maps had the same base directory which may break existing users, or should I create a 3rd base directory for secrets and config maps from projected volumes?

Comment (mikebryant) [reply pkg/checkpoint/pod.go]: @ericchiang Any update on this?

Comment (blerko) [reply pkg/checkpoint/pod.go]: @ericchiang sorry to bump this again, have you been able to take a look at this suggested solution?

Comment (rubencabrera) [reply pkg/checkpoint/pod.go]: @ericchiang @rphillips Is there anything to be done to progress on this PR?

Comment (rphillips) [reply pkg/checkpoint/pod.go]: Is there a reason why this code path is restricted to only all secrets or all configmaps? It feels like we can make this less error prone and support any combination of projections.

I worry about the edge cases and currently running deployments breaking. Adding a flag to enable projection (`--with-projection`) would mitigate the concern so we can test in current deployments. 

/cc @aaronlevy thoughts?

Comment (aaronlevy) [reply pkg/checkpoint/pod.go]: I would lean toward not adding flags that changes behavior. If we can't reason about safety - adding a flag is just going to make reasoning about safety of the behavior even more difficult in the future.

Comment (stuart-warren) [reply pkg/checkpoint/pod.go]: @rphillips we could create a separate base directory to store projected volume files and would have made that change had @ericchiang not made a comment that they needed to think about it a little

Comment (blerko) [reply pkg/checkpoint/pod.go]: @rphillips @aaronlevy Do you have any further thoughts on this?

**Comment by rphillips** [comment]
/ok-to-test
coreosbot run e2e checkpointer

**Comment by rphillips** [comment]
coreosbot run e2e checkpointer

**Comment by rphillips** [comment]
third time is the charm.

coreosbot run e2e checkpointer

**Comment by rphillips** [comment]
@aaronlevy any objections to merging this? Looks good to me and passes tests.

File: pkg/checkpoint/config_map.go [hunk 4]
```diff
@@ -50,13 +52,70 @@ func (c *checkpointer) checkpointConfigMap(namespace, podName, configMapName str
...
+	}
+	if err != nil {
+		return fmt.Errorf("failed to retrieve configMap %s/%s: %v", namespace, configMapName, err)
+	}
+
+	basePath := configMapPath(namespace, podName, volumeName)
+	if err = os.MkdirAll(basePath, 0700); err != nil {
+		return fmt.Errorf("failed to create configMap checkpoint path %s: %v", basePath, err)
+	}
+	if err = os.Chown(basePath, uid, gid); err != nil {
```
Comment (aaronlevy) [inline pkg/checkpoint/config_map.go]: nit: We're potentially creating parent directories via `MkdirAll` call above, but this would only set ownership on the full path.

File: pkg/checkpoint/pod.go (hunk 3 above)
Comment (aaronlevy) [reply pkg/checkpoint/pod.go]: Can we add tests for all projected volume types (as well as mixed types within same pod) to demonstrate that this has been resolved?

File: pkg/checkpoint/secret.go [hunk 5]
```diff
@@ -50,14 +53,71 @@ func (c *checkpointer) checkpointSecret(namespace, podName, secretName string, u
...
+	}
+	if err != nil {
+		return fmt.Errorf("failed to retrieve secret %s/%s: %v", namespace, secretName, err)
+	}
+
+	basePath := secretPath(namespace, podName, volumeName)
+	if err = os.MkdirAll(basePath, 0700); err != nil {
+		return fmt.Errorf("failed to create secret checkpoint path %s: %v", basePath, err)
+	}
+	if err = os.Chown(basePath, uid, gid); err != nil {
```
Comment (aaronlevy) [inline pkg/checkpoint/secret.go]: Same nit as above: only setting ownership on full path, but potentially creating parent dirs.

File: pkg/checkpoint/secret.go [hunk 6]
```diff
@@ -50,14 +53,71 @@ func (c *checkpointer) checkpointSecret(namespace, podName, secretName string, u
...
+// The path to the secret data becomes: checkpointSecretPath/namespace/podname/volumename/secret.file
+// Where each "secret.file" is a path from the SecretProjection.KeyToPath field.
+func (c *checkpointer) checkpointSecretProjection(namespace, podName, volumeName string, secretProjection *v1.SecretProjection, uid, gid int) error {
+
+	secretName := secretProjection.Name
+	isOptional := secretProjection.Optional != nil && *secretProjection.Optional
+
+	secret, err := c.apiserver.Core().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil && isOptional {
+		return nil
```
Comment (aaronlevy) [inline pkg/checkpoint/secret.go]: Same nit as above: potentially hiding errors we should surface

File: pkg/checkpoint/secret.go [hunk 7]
```diff
@@ -50,14 +53,71 @@ func (c *checkpointer) checkpointSecret(namespace, podName, secretName string, u
...
 		}
 	}
 
 	return basePath, nil
 }
 
+// checkpointSecretProjection will locally store secret data from volumes with ProjectedVolumeSources.
+// The path to the secret data becomes: checkpointSecretPath/namespace/podname/volumename/secret.file
+// Where each "secret.file" is a path from the SecretProjection.KeyToPath field.
+func (c *checkpointer) checkpointSecretProjection(namespace, podName, volumeName string, secretProjection *v1.SecretProjection, uid, gid int) error {
```
Comment (aaronlevy) [inline pkg/checkpoint/secret.go]: nit: This seems to duplicate a fair amount of behavior with checkpointConfigMapProjection (on a quick scan differences are the checkpoint path and retrieving the object from api). Might be able to merge those behaviors with something like a `checkpointProjection` that just takes one of the object types.

File: pkg/checkpoint/projected.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+				}
+			}
+			if volumeHasConfigMapProjection(v) {
+				err := c.checkpointProjectedConfigMaps(pod.Namespace, pod.Name, v.Name, p.Sources, uid, gid)
+				if err != nil {
+					return nil, fmt.Errorf("failed to checkpoint projected configmaps for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+				}
+			}
+			if volumeHasDownwardAPIProjection(v) {
+				return nil, fmt.Errorf("no support to checkpoint projected DownwardAPI for pod %s/%s", pod.Namespace, pod.Name)
```
Comment (aaronlevy) [inline pkg/checkpoint/projected.go]: Not supporting all types was brought up as a concern. Are we okay with shipping without support for this type? At least some support is better than none, but want to make sure other's okay with this.

cc @dghubble

File: pkg/checkpoint/pod.go (hunk 3 above)
Comment (aaronlevy) [reply pkg/checkpoint/pod.go]: Also was cleanup of the checkpointed data handled?

File: pkg/checkpoint/projected.go (hunk 8 above)
Comment (dghubble) [reply pkg/checkpoint/projected.go]: The not supporting downward API projection part sounds fine at least, since it wasn't supported before. If folks using kubeadm or other custom tools use manifests with downwardapi projections, they could contemplate how to implement it.

File: pkg/checkpoint/pod_test.go [hunk 9]
```diff
@@ -170,6 +170,114 @@ func TestSanitizeCheckpointPod(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "podname",
+					Namespace: "podnamespace",
+				},
+				Spec: v1.PodSpec{
+					Volumes: []v1.Volume{{
+						Name: "k8s-certs",
+						VolumeSource: v1.VolumeSource{
+							Projected: &v1.ProjectedVolumeSource{
+								Sources: []v1.VolumeProjection{
```
Comment (dghubble) [inline pkg/checkpoint/pod_test.go]: Is there a test with a ConfigMap VolumeProjection I'm missing?
